pysrc/bytewax/inputs.py

```diff
@@ -0,0 +1,204 @@
+    ...     in order(
+    ...         items,
+    ...         datetime.timedelta(seconds=2),
+    ...         itemgetter("timestamp"),
+    ...     )
+    ... ]
+    ['b', 'c']
+
+    Args:
+    
+        wrap_iter: Existing input iterable.
+   
+        window_length: Maximum out-of-order-ness before dropping
+            values.
+
+        time_getter: Function to call to produce a timestamp for each
+            value. Defaults to using the first item in each value;
+            e.g. `(time, x)`.
+
+        on_drop: Function to call with each dropped item. Defaults to
+            nothing.
+
+    Yields: Values in order.
+    """
+    buffer_heap = []
+    buffer_close_time = None
+    for item in wrap_iter:
+        time = time_getter(item)
+        if buffer_close_time is None or time >= buffer_close_time:
+            heapq.heappush(buffer_heap, (time, item))
+
+            buffer_close_time_item = time - window_length
+            if buffer_close_time is None or buffer_close_time_item > buffer_close_time:
+                buffer_close_time = buffer_close_time_item
+        
+                while len(buffer_heap) > 0 and buffer_heap[0][0] < buffer_close_time:
+                    time, item = heapq.heappop(buffer_heap)
+                    yield item
+            # Else: This item didn't push forward the window, so no
+            # items could be emitted.
+        elif on_drop:
+            on_drop(item)
+
+    # Emit everything remaining.
+    while len(buffer_heap) > 0:
+        time, item = heapq.heappop(buffer_heap)
+        yield item
+        
+
+def tumbling_epoch(
+        wrap_iter: Iterable,
+        epoch_length: Any,
+        time_getter: Callable[[Any], Any] = lambda _: datetime.datetime.now(),
+        epoch_0_start_time: Any = None,
+) -> Iterable[Tuple[int, Any]]:
+    """All inputs within a tumbling window are part of the same epoch.
+
+    The time of the first item will be used as start of the 0
+    epoch. Out-of-order items will cause issues as Bytewax requires
+    inputs to dataflows to be in epoch order. See `order()`.
+
+    >>> items = [
+    ...     {
+    ...         "timestamp": datetime.datetime(2022, 2, 22, 1, 2, 3),
+    ...         "value": "a",
+    ...     },
+    ...     {
+    ...         "timestamp": datetime.datetime(2022, 2, 22, 1, 2, 4),
+    ...         "value": "b",
+    ...     },
+    ...     {
+    ...         "timestamp": datetime.datetime(2022, 2, 22, 1, 2, 8),
+    ...         "value": "c",
+    ...     },
+    ... ]
+    >>> [
+    ...     (epoch, item["value"])
+    ...     for epoch, item
+    ...     in tumbling_epoch(
+    ...         items,
+    ...         datetime.timedelta(seconds=2),
+    ...         itemgetter("timestamp"),
+    ...     )
+    ... ]
+    [(0, 'a'), (0, 'b'), (2, 'c')]
+
+    By default, uses "ingestion time" and you don't need to specify a
+    way to access the timestamp in each item.
+
+    >>> import pytest; pytest.skip("Figure out sleep in test.")
```

Comment:
Should this be here?

Reply:
I've been trying to keep all the doctests executable without errors, but this one doesn't actually work because of the wall time stuff. So this marks it as "skip" so we know to fix it later.